Support manager address changes

aaronlehmann opened the pull request:
One of the things necessary to unblock full support for swarm mode in Docker for Mac/Windows is support for dynamic IP addresses on managers. Users of Docker for Mac/Windows are much more likely to be running in an environment like a laptop with WiFi. Of course, this will also be useful in other setups.

This PR begins to explore allowing address changes at the raft level. Each manager keeps track of the last source address it saw in RPCs from each other manager. When a send fails, and the most recently seen source address differs from the address we're sending to, and the observed address passes a health check, we switch to that address. Also, if this happens on the leader, the new address will be written to raft. Nodes are protected from election loops during an address change by ignoring vote requests unless the sender passes a health check.

One nice property of this approach is that address changes are recognized automatically, and don't need the admin or the manager whose address changed to specifically request an address change.

I haven't tested this yet, so the next step will be to write a unit test that confirms it works.

There will probably need to be libnetwork changes to fully support dynamic addresses.

Also, this does not address the case where quorum is lost (i.e. all nodes restarting and changing their addresses at the same time). We will probably need some kind of variant of `--force-new-cluster` to deal with that.

cc @justincormack @LK4D4

LK4D4 commented on manager/state/raft/raft.go:
```diff
@@ -1178,21 +1239,64 @@ func (n *Node) sendToMember(members map[uint64]*membership.Member, m raftpb.Mess
+		return
+	}
+
+	healthClient := api.NewHealthClient(member.Conn)
```
Maybe method `member.HealthCheck(ctx)`?

LK4D4 commented on manager/state/raft/raft.go:
```diff
@@ -799,7 +830,37 @@ func (n *Node) ProcessRaftMessage(ctx context.Context, msg *api.ProcessRaftMessa
-	n.cluster.ReportActive(msg.Message.From)
+	// Reject vote requests from unreachable peers
+	if msg.Message.Type == raftpb.MsgVote {
+		member := n.cluster.GetMember(msg.Message.From)
```
Maybe check `cluster.Active` before healthCheck?

aaronlehmann replied:
Actually I don't think checking `Active` would work. If the member requesting the election is not the leader and the member that receives the request is not the leader either, the first member would be considered inactive (because it only communicates with the leader).

LK4D4 replied:
That's kinda sad. I thought about this yesterday, too.

LK4D4 commented on manager/state/raft/raft.go:
```diff
@@ -1178,21 +1239,64 @@ func (n *Node) sendToMember(members map[uint64]*membership.Member, m raftpb.Mess
 	}
 }
 
+func (n *Node) handleAddressChange(member *membership.Member, reconnectAddr string) {
```
should return error I think

aaronlehmann replied:
It's called from a goroutine as the last thing the goroutine does, so I'm not sure it's useful to return an error. It logs errors instead.

LK4D4 replied:
I think it's not "scalable" to write such functions. What if you'll want to use it in another place? Or write a test for it? I think(and with errors package, it's even cooler) it's better to return errors and log in one place.

LK4D4 commented on manager/state/raft/raft.go:
```diff
@@ -1178,21 +1239,64 @@ func (n *Node) sendToMember(members map[uint64]*membership.Member, m raftpb.Mess
+
+	// If we're the leader, write the address change to raft
+	updateCtx, cancelUpdate := context.WithTimeout(n.Ctx, time.Duration(n.Config.ElectionTick)*time.Second)
+	n.updateMember(updateCtx, reconnectAddr, member.RaftID, member.NodeID)
```
we should at least log error from here

aaronlehmann replied:
Agreed

LK4D4 commented on manager/state/raft/raft.go:
```diff
@@ -1178,21 +1239,64 @@ func (n *Node) sendToMember(members map[uint64]*membership.Member, m raftpb.Mess
+
+	healthClient := api.NewHealthClient(member.Conn)
+
+	healthCtx, cancelHealth := context.WithTimeout(n.Ctx, time.Duration(n.Config.ElectionTick)*time.Second)
```
btw, it bothered me for a long time, that `TickInterval` is configurable, but not exposed inside `Node` object, it should be something like `n.tickInterval` not `time.Second` I think.

LK4D4 commented on manager/state/raft/raft.go:
```diff
@@ -1178,21 +1239,64 @@ func (n *Node) sendToMember(members map[uint64]*membership.Member, m raftpb.Mess
-			}
+			return
+		}
+		if n.cluster.ReplaceMemberConnection(m.To, conn, newConn, conn.Addr, false) != nil {
```
not a fan of such syntax

LK4D4 replied:
Also error probably should be logged.

aaronlehmann replied:
The error's basically impossible. It can only happen if the node is removed while running this function.

Any suggestions on better syntax?

LK4D4 replied:
I'd like to see `if err := ...; err != nil`
If it's impossible - why check it at all? Also, I don't see anything preventing node removal while this function is run.

LK4D4 commented on manager/state/raft/membership/cluster.go:
```diff
@@ -180,7 +181,7 @@ func (c *Cluster) ReplaceMemberConnection(id uint64, oldConn *Member, newConn *M
 	}
 
-	if oldConn.Conn != oldMember.Conn {
+	if !force && oldConn.Conn != oldMember.Conn {
```
When conn can be the same and Addr different?

aaronlehmann replied:
The normal use case for this function is to bounce the connection. I'm extending it here to also allow a change of address. Maybe it would be less confusing to have a different function for address changes.

LK4D4 replied:
IMO conn and addr should be always in sync. It's quite confusing, otherwise.

aaronlehmann replied:
They are always in sync. In the case where the address doesn't change, the old address is passed in, so setting it is a no-op.

LK4D4 replied:
Yeah, that's what I'm trying to understand. If they always in sync, then `(oldAddr != newAddr) == (oldConn != newConn)`. And what the point of `force`, then?

aaronlehmann replied:
This part is not related to the address. It's about replacing the GRPC connection with a new connection. Imagine the address never changed, but an RPC failed. Instead of depending on GRPC's slow backoff, we close that connection and initiate a new one. This function replaces the old connection with the new one in the `Cluster` object.

Looking at the code before these changes might make it clearer.

LK4D4 replied:
force was added in this pr, so it's somehow related to addr :) what the point of force argument?

aaronlehmann replied:
The old version of this function is kind of like a check-and-set. It won't replace the connection if it has already been replaced. This is to prevent multiple `sendToMember` goroutines from fighting each other. One could bounce the connection, then this kills the old connection, which causes another goroutine to have an RPC failure, which causes it to replace the connection again, and so on.

It's different when we see an address change in raft. Then we can't do best-effort. We need to replace the connection no matter what. So that's why there's a force option.

LK4D4 replied:
Ok. Thanks for explanation!

aaronlehmann commented:
Addressed review comments. Tried working on a unit test for this, but I don't see a good way. Both the bind address and source address for outgoing connections have to change. The unit tests just use 127.0.0.1.

LK4D4 commented on manager/state/raft/raft.go:
```diff
@@ -1178,21 +1235,68 @@ func (n *Node) sendToMember(members map[uint64]*membership.Member, m raftpb.Mess
+		return errors.Wrapf(err, "could connect to member ID %x at observed address %s", member.RaftID, reconnectAddr)
+	}
+
+	healthCtx, cancelHealth := context.WithTimeout(n.Ctx, time.Duration(n.Config.ElectionTick)*n.opts.TickInterval)
```
Hmmm. I'd propose maybe do `(*Node).MemberHealthCheck(m *Member)` and include context creation and error handling there entirely. It should reduce duplication.

aaronlehmann replied:
Hmm, in one case a custom timeout needs to be specified, so it would be `(*Node).MemberHealthCheck(m *Member, timeout time.Duration)`, and I think passing in a context is actually more elegant than passing in a timeout and having the function construct its own context.

For error handling, the error messages we produce are different in the various cases (for example `failed to check health of member %x after receiving vote request`, so I'd prefer to leave error handling to the callers.

LK4D4 replied:
@aaronlehmann Didn't think about custom timeouts, thanks! But error case should be handled by wrapping IMO. Caller mostly doesn't care if it was an error or unhealthy service, it just needs to add its context - i.e. "health check failed after a vote".

aaronlehmann replied:
@LK4D4: Updated, WDYT?

LK4D4 commented on manager/state/raft/raft.go:
```diff
@@ -1178,21 +1234,68 @@ func (n *Node) sendToMember(members map[uint64]*membership.Member, m raftpb.Mess
+	if err := n.updateMember(updateCtx, reconnectAddr, member.RaftID, member.NodeID); err != nil {
+		return errors.Wrap(err, "failed to update member address in raft")
+	}
+	cancelUpdate()
```
should be cancelled in defer

aaronlehmann replied:
Noticed this, forgot to fix it, thanks. Fixed now.

LK4D4 commented on manager/state/raft/raft.go:
```diff
@@ -1387,6 +1492,40 @@ func (n *Node) applyAddNode(cc raftpb.ConfChange) error {
+		return ErrMemberUnknown
+	}
+	if oldMember.NodeID != newMember.NodeID {
+		return errors.New("node ID does not match on node update")
```
can we have log message here with IDs of nodes?

aaronlehmann replied:
Added, though I don't expect this sanity check to ever trigger.

LK4D4 commented on manager/state/raft/raft.go:
```diff
@@ -1387,6 +1492,40 @@ func (n *Node) applyAddNode(cc raftpb.ConfChange) error {
+	if err != nil {
+		return errors.Errorf("could connect to member ID %x at %s: %v", newMember.RaftID, newMember.Addr, err)
+	}
+	if n.cluster.ReplaceMemberConnection(newMember.RaftID, oldMember, newConn, newMember.Addr, true) != nil {
```
please use `if err := ...; err != nil` syntax here to indicate that it's error. Also, I think it's probably ok to return error here. It would be really hard to catch this race, which is quite possible.

aaronlehmann commented:
Feedback addressed. Also squashed the commits, because new changes made it hard to maintain the old commit structure.

I tested this with 3 VMs. I set up a 3-manager swarm, then I restarted one of the follower VMs, which caused it to get a new IP address. When it came back up, it was able to participate in the swarm. I also restarted the leader's Docker engine a few times to trigger leader changes, and confirmed that things work when the node with the changed address becomes the leader. I also confirmed that restarting all engines at once works after the address change.

I'm also going to test adding new nodes to the cluster after the address change, to make sure they pick up the correct address and not the original one.

aaronlehmann commented:
Adding a new node after the address change works. No problems when the new node becomes the leader, or when the one whose address changed becomes the leader.

aaronlehmann commented:
Docker integration tests are passing with this - except for `TestSwarmTaskListFilter`, which is also failing on docker master for me (strange, CI seems fine).

LK4D4 commented on manager/state/raft/raft.go:
```diff
@@ -1178,21 +1225,64 @@ func (n *Node) sendToMember(members map[uint64]*membership.Member, m raftpb.Mess
+	defer cancelHealth()
+
+	if err := newConn.HealthCheck(healthCtx); err != nil {
+		return errors.Errorf("%x failed health check at observed address %s", member.RaftID, reconnectAddr)
```
Kinda cool, but you should wrap original error here as well.

LK4D4 commented on manager/state/raft/raft.go:
```diff
@@ -799,7 +823,30 @@ func (n *Node) ProcessRaftMessage(ctx context.Context, msg *api.ProcessRaftMessa
+
+		if err := member.HealthCheck(healthCtx); err != nil {
+			n.Config.Logger.Warningf("member %x which sent vote request failed health check", msg.Message.From)
+			return nil, ErrMemberUnreachable
```
This is cool until we return this error from one place, but even now it would be hard to sync logs from one node to another. I think it's better to return full error here with Wrap and stuff.

aaronlehmann replied:
@LK4D4: Sorry, I don't understand why it's useful to wrap the error. You would get `member %x which sent vote request failed health check: health check failed`. This would just add redundancy.

LK4D4 replied:
It would be "health check failed" or grpc error from health check. Also if I later return ErrMemberUnreachable somewhere else, it will be harder to debug IMO.

aaronlehmann replied:
@LK4D4: Updated, PTAL.

LK4D4 commented:
LGTM
ping @aluzzardi @dongluochen @nishanttotla

LK4D4 commented:
Going to merge this, because nobody appeared to review it.
